dadguide/monster_index.py

```diff
@@ -1,34 +1,155 @@
-        self.index = defaultdict(list)
+        self.index = defaultdict(set)
         self._build_index()
 
+    def update(self, other: "Mapping[Any, set]"):
+        for k,v in other.items():
+            self.index[k].update(v)
+
     def _build_index(self):
-        self.index = defaultdict(list)
-        self.monster_name_index = self._build_monster_name_index(self.monsters)
-        self.series_name_index = self._build_series_name_index(self.series)
+        self.index = defaultdict(set)
+        self.update(self._build_monster_name_index(self.monsters))
+        self.update(self._build_color_index(self.monsters))
+        self.update(self._build_better_series_index(self.monsters))
+        self.update(self._build_evo_index(self.monsters))
+        self.update(self._build_etc_index(self.monsters))
+        #self.series_name_index = self._build_series_name_index(self.series)
 
     def _build_monster_name_index(self, monsters):
-        tokens = defaultdict(list)
+        tokens = defaultdict(set)
         for monster in monsters:
             for token in self._name_to_tokens(monster.name_en):
-                tokens[token.lower()].append(monster)
+                tokens[token.lower()].add(monster)
         return tokens
 
     def _build_series_name_index(self, series):
-        tokens = defaultdict(list)
+        tokens = defaultdict(set)
         for s in series:
             for token in self._name_to_tokens(s.name_en):
-                tokens[token.lower()].append(s)
+                tokens[token.lower()].add(s)
         return tokens
 
     def _name_to_tokens(self, name):
         name = re.sub(r'[\-+]', ' ', name.lower())
         name = re.sub(r'[^a-z0-9 ]', '', name)
         return [t.strip() for t in name.split() if t]
+
+    def _build_color_index(self, monsters):
+        tokens = defaultdict(set)
+
+        # Main Color
+        for c in COLOR_MAP:
+            l = {m for m in monsters if m.attr1.value == c.value}
+            for t in COLOR_MAP[c]:
+                tokens[t].update(l)
+
+        # Sub Color
+        for c1 in COLOR_MAP:
+            for c2 in COLOR_MAP:
+                l = {m for m in monsters if m.attr1.value == c1.value and m.attr1.value == c2.value}
+                for t1 in COLOR_MAP[c1]:
+                    for t2 in COLOR_MAP[c2]:
+                        tokens[t1+t2].update(l)
+                        tokens[t1+"/"+t2].update(l)
+
+        return tokens
+
+    def _build_better_series_index(self, monsters):
+        tokens = defaultdict(set)
+
+        # Series
+        for s in SERIES_MAP:
+            l = {m for m in monsters if m.series_id == s}
+            for t in SERIES_MAP[s]:
+                tokens[t].update(l)
+
+        return tokens
+
+    def _build_evo_index(self, monsters):
+        tokens = defaultdict(set)
+
+        # Base
+        # NOTHING HERE.  THIS IS A SPECIAL CASE
+
+        # Evo
+        l = {m for m in monsters if m.cur_evo_type.value == 1}
```

Comment:
you're looping over monsters too many times in this function. loop over it once, and use if statements. (this is when list comprehensions suck)